src/QsCompiler/Transformations/CallGraph/CallGraphWalker.cs

```diff
@@ -0,0 +1,507 @@
+        /// those callables that the entry points depend on will be included in the graph.
+        /// </summary>
+        public static void PopulateTrimmedGraph(SimpleGraphBuilder graph, QsCompilation compilation) => SimpleCallGraphWalker.PopulateTrimmedGraph(graph, compilation);
+
+        /// <summary>
+        /// Populates the given graph based on the given compilation. Only the compilation's entry points and
+        /// those callables that the entry points depend on will be included in the graph.
+        /// </summary>
+        public static void PopulateConcreteGraph(ConcreteGraphBuilder graph, QsCompilation compilation) => ConcreteCallGraphWalker.PopulateConcreteGraph(graph, compilation);
+
+        private static class BaseCallGraphWalker<TGraph, TNode, TEdge>
+            where TGraph : CallGraphBuilder<TNode, TEdge>
+            where TNode : CallGraphNodeBase
+            where TEdge : CallGraphEdgeBase
+        {
+            public abstract class TransformationState
+            {
+                internal TNode CurrentNode;
+                internal readonly TGraph Graph;
+
+                // The type parameter resolutions of the current expression.
+                internal IEnumerable<TypeParameterResolutions> ExpTypeParamResolutions = new List<TypeParameterResolutions>();
+                internal QsNullable<Position> CurrentStatementOffset;
+                internal QsNullable<Range> CurrentExpressionRange;
+                internal readonly Stack<TNode> RequestStack = new Stack<TNode>(); // Used to keep track of the nodes that still need to be walked by the walker.
+                internal readonly HashSet<TNode> ResolvedNodeSet = new HashSet<TNode>(); // Used to keep track of the nodes that have already been walked by the walker.
+
+                internal TransformationState(TGraph graph)
+                {
+                    this.Graph = graph;
+                }
+
+                /// <summary>
+                /// Adds dependency to the graph from the current callable to the callable referenced by the given identifier.
+                /// </summary>
+                internal abstract void AddDependency(QsQualifiedName identifier);
+            }
+
+            public class StatementWalker<TState> : StatementTransformation<TState>
+                where TState : TransformationState
+            {
+                public StatementWalker(SyntaxTreeTransformation<TState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override QsStatement OnStatement(QsStatement stm)
+                {
+                    this.SharedState.CurrentStatementOffset = stm.Location.IsValue
+                        ? QsNullable<Position>.NewValue(stm.Location.Item.Offset)
+                        : QsNullable<Position>.Null;
+                    return base.OnStatement(stm);
+                }
```

Comment:
This looks like a recurring need in transformations: to keep track of the offset of the last seen statement in order to add it to an expression range. Capability inference needed to write a statement transformation for this as well. Maybe there is some common infrastructure we could make for this? @bettinaheim

Reply:
Yes, that would likely make sense. Not as part of this change though.